src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -475,6 +494,51 @@
    (utils/subject-key-identifier
      ca-public-key false)])
 
+(schema/defn read-compile-masters
+    "Returns a list of compile masters from the specified file.
+     Note that this method is agnostic of the file contents which could either
+     be node names (FQDNs) or certificate serial numbers so long as they are
+     organized as one item per line.
+     Likewise this method could/will be re-used"
+    [cm-file :- schema/Str]
+    (line-seq (io/reader cm-file)))
+
+(schema/defn generate-compile-master-serials
+  "Given a list of compile master nodes it will create a file containing 
+   serial numbers of their certificates (listed on separate lines).
+   Compile masters is expected have at least one entry (MoM)"
+   [cm-nodes :- [schema/Str]
+    compile-masters-serials-file :- schema/Str
+    signeddir :- schema/Str]
+
+   ;; TODO: We may want to account for the possibility of compile masters to
+   ;; be empty (by mistake, accident). In which case this can use the name
+   ;; of the current node as the MoM as a fallback
+  (log/debug (format "%s\n%s" (i18n/trs "generate-compile-master invoked ") compile-masters-serials-file))
+  (comment
+   (with-open [wtr (clojure.java.io/writer compile-masters-serials-file)]
+     (doseq [cm-node cm-nodes]
+       (let [cm-serial  (-> (path-to-cert signeddir cm-node)
+                         (utils/pem->cert)
+                         (utils/get-serial))]
+          (log/debug (format "%s: %3d\n" "Serial " cm-serial))
+          ;; These writes are failing with an error of "No matching method found: write for class java.io.BufferedWriter"
+          ;; The filename is valid and writable. Same code works in a test app
+          ;; (.write wtr cm-serial)
```

Comment:
It might not be able to  infer what cm-serial will be, and so can't chose the right method. Try adding a [type hint](https://clojure.org/reference/java_interop#typehints). Something like `(.write wtr ^String cm-serial)`.